### ISSUE TITLE
Invoke-WebRequest discards cookies on HTTP 30x redirects

### ISSUE THREAD

**Issue by nutjob4life** [PowerShell/PowerShell#6065]
I'm attempting to access a web service using `Invoke-WebRequest` that has the following interaction:

1. You first GET a URL, passing an HTTP Basic auth header.
2. The server passes back a 302 redirection with both a `Set-Cookie` and a `Location` header.
3. You then GET *that* URL from the `Location` header, passing in the `Cookie` you got in step 2.

Normally this works fine with `/usr/bin/curl` passing in the `--user`, `--cookie-jar`, and `--location` options. However, PowerShell's `Invoke-WebRequest` discards any cookies as soon as it sees the 302 redirection. I've tested this with PowerShell on Windows and macOS.

Steps to reproduce
------------------

```
$uri = "https://host/api"
$username = "username"
$password = "password"
$pair = "${username}:${password}"
$bytes = [System.Text.Encoding]::ASCII.GetBytes($pair)
$base64 = [System.Convert]::ToBase64String($bytes)
$basicAuth = "Basic $base64"
$headers = @{ Authorization = $basicAuth }
Invoke-WebRequest -Uri $uri -Headers $headers -OutFile "output.dat" -SessionVariable cookie_jar
```

Expected behavior
-----------------

(Expressed in terms of HTTP traffic.)

```
> GET /api HTTP/1.1
> Host: host
> Authorization: Basic base64dataHere===
> 
< HTTP/1.1 302 Found
< Set-Cookie: specialCookie=someOpaqueData123; Max-Age=3600; Path=/download; secure
< Location: https://host/download?parameter=2700fa40-15c4-46b5-9a4f-73ef534c1889
> GET /download?parameter=2700fa40-15c4-46b5-9a4f-73ef534c1889 HTTP/1.1
> Host: host
> Authorization: Basic base64dataHere===
> Cookie: specialCookie=someOpaqueData123
> 
< HTTP/1.1 200 OK
< Content-Disposition: attachment; filename="yourdata.dat"
< Content-Type: application/octet-stream
< Content-Length: 372959765
...
```
Actual behavior
---------------

```
> GET /api HTTP/1.1
> Host: host
> Authorization: Basic base64dataHere===
> 
< HTTP/1.1 302 Found
< Set-Cookie: specialCookie=someOpaqueData123; Max-Age=3600; Path=/download; secure
< Location: https://host/download?parameter=2700fa40-15c4-46b5-9a4f-73ef534c1889
> GET /download?parameter=2700fa40-15c4-46b5-9a4f-73ef534c1889 HTTP/1.1
> Host: host
> Authorization: Basic base64dataHere===
> 
< HTTP/1.1 401 Unauthorized
< Content-Type: text/plain
<
< specialCookie not found
```

Environment data
----------------

```
PSVersion                      6.0.1                                                                                   
PSEdition                      Core                                                                                    
GitCommitId                    v6.0.1                                                                                  
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                 
PSRemotingProtocolVersion      2.3                                                                                     
SerializationVersion           1.1.0.1                                                                                 
WSManStackVersion              3.0                                                                                     
```

**Comment by markekraus** [comment]
Hmm I'd have to craft an endpoint to reproduce this. I suspect it is due to our striping of authorization headers on redirect. Rather than letting HttpClient handle the redirect, we handle it manual to prevent authorization headers from being sent. i suspect that when we do that, somehow the cookies are not being set on the cookie container.

**Comment by markekraus** [comment]
After creating a custom endpoint, I am not able to reproduce this issue. 

Upon further checking the code, we are setting the `HttpClientHandler.CookieContainer` to the `WebSession.Cookies` every time we build the request.

https://github.com/PowerShell/PowerShell/blob/bd865280f36ea127414395d724cff1c778ddba73/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs#L898

I'm using this simple ASP.NET app https://github.com/markekraus/RedirectCookieAuth/

then in PowerShell 6.0.1:

```powershell
$uri = 'http://localhost:5000/api/Redirect'
$username = "username"
$password = "password"
$pair = "${username}:${password}"
$bytes = [System.Text.Encoding]::ASCII.GetBytes($pair)
$base64 = [System.Convert]::ToBase64String($bytes)
$basicAuth = "Basic $base64"
$headers = @{ Authorization = $basicAuth }
Invoke-WebRequest -Uri $uri -Headers $headers -OutFile "c:\temp\output.json" -SessionVariable cookie_jar
Get-Content "c:\temp\output.json" | ConvertFrom-Json | ConvertTo-Json
``` 

result:

```json
{
  "args": {},
  "origin": "::1",
  "headers": {
    "Cookie": "special=test123",
    "Host": "localhost:5000",
    "Connection": "Keep-Alive",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Microsoft Windows 10.0.16299; en-US) PowerShell/6.0.1"
  },
  "method": "GET",
  "url": "http://localhost:5000/api/Get/"
}
```

Basically, `http://localhost:5000/api/Redirect` sets the `special` cookie to `test123` then redirects to `http://localhost:5000/api/Get/` which then spits out the details of the request, including the headers. You can see in the headers collection the `Cookie` request header shows the `special=test123` value.

**Comment by nutjob4life** [comment]
Thanks @markekraus. Something else must be going on in my environment.

**Comment by trir262** [comment]
Hello, I stumbled on this issue and can provide more context: if I submit a request with an Authentication header to a non IIS webserver, and the web server does not return a cookie, the subsequent request fails because the Authentication header is now missing, causing a http 401 (access denied).
A workaround is to use the MaximumRedirection option: 
`invoke-webrequest -method get -uri "-host_stripped-" -MaximumRedirection 0 -erroraction silentlycontinue -errorvariable redirecterror` 
and use the Location value to create a new request (with Authentication header) until I no longer receive HTTP 301 or HTTP 302 and then return the final result.

Here is an example of the requests on our server:
First request
GET https://-host_stripped-.com/api/device HTTP/1.1
Authorization: Basic -stripped-
User-Agent: Mozilla/5.0 (Windows NT; Windows NT 10.0; nl-BE) WindowsPowerShell/5.1.15063.936
Content-Type: application/json; charset=utf-8
Host: -host_stripped-.com

First response
HTTP/1.1 302 Found
Server: nginx
Date: Wed, 18 Apr 2018 22:44:35 GMT
Content-Type: application/json
Connection: keep-alive
Location: /api/device?limit=100
Cache-Control: private, no-cache, must-revalidate, no-store, max-age=0, post-check=0, pre-check=0
X-status-message: device index requires a limit
X-status-code: FOUND
Vary: Accept-Encoding
Strict-Transport-Security: max-age=157680000
Content-Length: 3121
----json file---

Second request, append previous 'Location' value in uri
GET https://-host_stripped-.com/api/device?limit=100 HTTP/1.1
Authorization: Basic -stripped-
User-Agent: Mozilla/5.0 (Windows NT; Windows NT 10.0; nl-BE) WindowsPowerShell/5.1.15063.936
Content-Type: application/json; charset=utf-8
Host: -host_stripped_.com

Second response
HTTP/1.1 200 OK
Server: nginx
Date: Wed, 18 Apr 2018 23:02:11 GMT
Content-Type: application/json
Connection: keep-alive
Cache-Control: private, no-cache, must-revalidate, no-store, max-age=0, post-check=0, pre-check=0
X-status-message: device index found with 5052 resources.
X-status-code: OK
Vary: Accept-Encoding
Strict-Transport-Security: max-age=157680000
Content-Length: 9290
--json file---

**Comment by markekraus** [comment]
@trir262 When the web endpoint redirects and you supply and authorization header, you need to use the `-PreserveAuthorizationOnRedirect` switch or else the Authorization header will be stripped as a security measure.

Working example:

```powershell
Invoke-WebRequest -Uri 'https://httpbin.org/redirect-to?url=https://httpbin.org/basic-auth/user/passwd' -Header @{Authorization = 'Basic dXNlcjpwYXNzd2Q=' } -PreserveAuthorizationOnRedirect
```

This 302 redirects to an authentication endpoint and preserves the Authorization header. Issuing the same command without `-PreserveAuthorizationOnRedirect` will result in a 401 response.



[More Info.](https://get-powershellblog.blogspot.com/2017/12/powershell-core-web-cmdlets-in-depth_24.html#L17)

**Comment by trir262** [comment]
Thank you Mark! This option is indeed what I was looking for. Unfortunately 5.1 does not have this.
Apologies for the confusion.

**Comment by markekraus** [comment]
@trir262 Correct. This is a new feature added in PowerShell 6.0.0. If this is an impacting issue for you in 5.1, I would suggest opening a [UserVoice](https://windowsserver.uservoice.com/forums/301869-powershell) or perhaps voting on an existing one.
